### ISSUE TITLE
[PR #152] xxx.sub == xxx.com now

### ISSUE THREAD

**Issue by drawkula** [MockbaTheBorg/RunCPM#154]
Playing with PR #152:
```
C1>dir myhi*.*
C: MYHI     C    |  MYHI     COM  |  MYHI     SUB
C1>type myhi.c
#include "stdio.h"
main()
{
        puts("No meltdown, no specte.");
        puts("No rowhammer, no brainslammer!");
        puts("\nCP/M!");
        puts(";-)");
}

C1>myhi.sub
No meltdown, no specte.
No rowhammer, no brainslammer!

CP/M!
;-)


C1>type myhi.sub
; create myhi.com
c myhi.c
myhi
;

C1>█
```
`C1>myhi.sub` seems to have executed the `.COM` file.
Is that intended and I just don't understand something?

**Comment by drawkula** [comment]
It should have compiled the source by `c myhi.c` which would not have been silent, before starting the resulting COM file if it would have called the subfile correctly.
That would look like:
```
C1>do myhi
SuperSUB V1.1


C1$; CREATE MYHI.COM
C1$C MYHI.C
Hi-Tech Z80 C Compiler (CP/M-80) V3.09-16
Copyright (C) 1984-87 HI-TECH SOFTWARE
Updated from https://github.com/agn453/HI-TECH-Z80-C


C1$MYHI
No meltdown, no specte.
No rowhammer, no brainslammer!

CP/M!
;-)


C1$;
C1>▙
```
After removing `myhi.com` the extension-less command is interpreted as subfile:
```
C1>era myhi.com
C1>myhi


C1$; CREATE MYHI.COM
C1$C MYHI.C
Hi-Tech Z80 C Compiler (CP/M-80) V3.09-16
Copyright (C) 1984-87 HI-TECH SOFTWARE
Updated from https://github.com/agn453/HI-TECH-Z80-C


C1$MYHI
No meltdown, no specte.
No rowhammer, no brainslammer!

CP/M!
;-)


C1$;
C1>█
```
Same with `.SUB` extension if the COM file is not there:
```
C1>era myhi.com
C1>myhi.sub


C1$; CREATE MYHI.COM
C1$C MYHI.C
Hi-Tech Z80 C Compiler (CP/M-80) V3.09-16
Copyright (C) 1984-87 HI-TECH SOFTWARE
Updated from https://github.com/agn453/HI-TECH-Z80-C


C1$MYHI
No meltdown, no specte.
No rowhammer, no brainslammer!

CP/M!
;-)


C1$;
C1>█
```
Now `myhi.com` and `myhi.sub` exist:
```
C1>myhi.sub
No meltdown, no specte.
No rowhammer, no brainslammer!

CP/M!
;-)


C1>█
```
No sign of compiler output.

**Comment by geowar1** [comment]
What’s the behavior using SUBMIT command (with & without my PR)?

**Comment by MockbaTheBorg** [comment]
I see Drawkula's point. He explicitly called MyHi.SUB, but this executes the .COM. We probably shouldn't be overriding the file extension without checking first.

**Comment by geowar1** [comment]
Ah! I think I got it… Yeah, I’ll check the extension first. Good catch. ;-)
PR #152 is merged so I’ll submit a new PR ASAP. ;-)

**Comment by MockbaTheBorg** [comment]
Question is: how would CP/M 3 behave in this case. We should then emulate the behavior as close as we can.

**Comment by geowar1** [comment]
IIRC vanilla CP/M 3 requires the “SUBMIT” command. ZCPR3 added the auto-submit support.

**Comment by geowar1** [comment]
Normally I don’t specify a file type when typing a command… Are there other types supported (at the beginning of the command prompt) other than .COM?

**Comment by geowar1** [comment]
The old (pre-PR #152) behavior always set the file type to .COM.
I think the correct fix would be to only override the file type if it’s blank.

**Comment by geowar1** [comment]
Note: That would mean disabling the auto-search (default drive, user zero; drive a, current user; drive A, user 0)… also (if the file type wasn’t blank).

**Comment by geowar1** [comment]
Ok, new PR (#155) submitted.

**Comment by geowar1** [comment]
Note: I also discovered a bug (I introduced in PR #152) that wouldn’t let the DU: command switch back to drive A:. Fixed(?) it also. Please test! ;-)

**Comment by MockbaTheBorg** [comment]
@drawkula should be working as expected now. Please check.

also ... if you could do some regression test we would be grateful.

**Comment by drawkula** [comment]
So far the only other thing I found is the args being off by one:
```
C1>type subargs.sub
; 1:$1 2:$2

C1>submit subargs a b c


C1$; 1:A 2:B
C1>subargs a b c


C1$; 1:B 2:C
C1>█
```

**Comment by MockbaTheBorg** [comment]
I think the .COM over .SUB part is resolved. Will check the parameters.

**Comment by MockbaTheBorg** [comment]
@drawkula can you check now?
```
B0>ed args.sub

NEW FILE
     : *i
    1:  ; 1:$1 2:$2 3:$3
    2:  →
     : *e


RunCPM Version 5.8 (CP/M 2.2 60K)

B0>submit args.sub a b c


RunCPM Version 5.8 (CP/M 2.2 60K)

B0$; 1:A 2:B 3:C
B0>args.sub a b c


RunCPM Version 5.8 (CP/M 2.2 60K)

B0$; 1:A 2:B 3:C

**Comment by drawkula** [comment]
```
C1>subargs one two three


C1$; 1:ONE 2:TWO
```
Looks ok.  Thanks!

Things like `c $1.c` in submit files just are needed.  E.g. as compile and run script:
```
C1>type cr.sub
era $1.com
c $1.c
$1

C1>cr hchs


C1$ERA HCHS.COM
C1$C HCHS.C
Hi-Tech Z80 C Compiler (CP/M-80) V3.09-16
Copyright (C) 1984-87 HI-TECH SOFTWARE
Updated from https://github.com/agn453/HI-TECH-Z80-C


C1$HCHS
Hello, CP/M!
Hello, StdIO!


C1>_
```

**Comment by geowar1** [comment]
So… can this issue be closed?
